### DISCUSSION TITLE
PixelsOut support for Waveshare 400x300, 4.2inch E-Ink display module

### DISCUSSION THREAD

**Discussion by ivandrosero** [Moddable-OpenSource/moddable#1438] [Q&A]
Hello, 

I am using the following ESP32-WROOM based device, integrated with a Waveshare's 4.2 inch e-ink display:

- https://docs.paperd.ink/docs/hardware/hardware/
- https://www.waveshare.com/4.2inch-e-paper-module.htm

I have working code using micropython, but I would love to switch to moddable SDK.  Is PixelsOut API an option? 

Pin-Out information below.   

![image](https://github.com/user-attachments/assets/7246a961-57bd-4063-b95f-0bc828b45a23)

![image](https://github.com/user-attachments/assets/9be02730-d01a-4489-859e-febff19ef28e)

SPI Pins
- SCK: GPIO18
- MOSI: GPIO23
- MISO: GPIO19

E-paper Display Pins
- EPD CS: GPIO22
- EPD DC: GPIO15
- EPD BUSY: GPIO34
- EPD RESET: GPIO13
- EPD Enable: GPIO12

Thank you very much for your time and attention.

Ivan

**Comment by phoddie** [comment]
Welcome, Ivan, and thank you for the suggestion. I'm a big fan of ePaper displays and the Moddable SDK has extensive support for them (check out [this blog post](https://moddable.com/blog/epaper/)).

It is surely possible to to support the Waveshare display. It looks like a really nice display - big and a couple shades of gray in addition to black and white. In my experience, we end up writing a new display driver for each E-Ink display, because the drivers are often pretty different. The [Arduino driver code](https://www.waveshare.com/wiki/4.2inch_e-Paper_Module_Manual#Working_With_Arduino) for this display looks reasonably straightforward, but it is 600 lines of code to understand.

Do you have any experience writing video drivers? If so, I'd be happy to help guide you through bringing this one up.

**Reply by ivandrosero** [reply]
Hello,

I do not have experience writing one, but happy to give it a try translating to javascript.  Looks like in this Arduino project

https://github.com/waveshareteam/e-Paper/blob/master/Arduino/epd4in2_V2/epd4in2_V2.ino

the place to look would be in the implementations of these headers: 

    #include "epd4in2_V2.h"
    #include "imagedata.h"
    #include "epdpaint.h"
?

Not a C/C++ programmer, but if you can tell me where to even put this stuff so that the Moddable build tools pick it up, and I can start debugging, that would already be more of a start than I currently understand.

Or, an it be done entirely in javascript?  

Thanks for your tips here, much appreciated!

**Comment by phoddie** [comment]
Hey, apologies for the delayed response. The holidays distracted me.

You should be able to implement the entire thing in JavaScript. In fact, there is a 100% JavaScript ePaper driver in the repository already. The [gdew0154m09](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/gdew0154m09/gdew0154m09.js) for the M5Core Ink device shows how to do it. The entire is under 400 lines of code, including all the communication with the display controller and a PixelsOut for integration with Commodetto. It buffers the entire image in memory, which is entirely practical for ePaper. It is a 1-bit driver, so it will be some work to support the 2-bit mode of your display, but you could start with 1-bit and go from there.

FYI – there's also the [IT8951](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/it8951/it8951.js) driver for the M5 Paper display which uses 4-bit gray pixels. It has a similar structure to the gdew0154m09 driver. It is entirely JavaScript except for one optional function to apply a simple filter. That may be a good starting point for 2-bit gray, since you could render to 4-bit gray (supported natively by Poco, 2-bit gray is not) and then downsample to 2-bit gray when outputting to the display controller.
